File: 백준/Gold/1261. 알고스팟/알고스팟.py
import sys
from collections import deque

input = sys.stdin.readline

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

def bfs():
    visit[0][0] = 0
    queue = deque()
    queue.append((0, 0))

    while queue:
        x, y = queue.popleft()
        if x == m - 1 and y == n - 1:
            return visit[x][y]
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < m and 0 <= ny < n and visit[nx][ny] == -1:
                if maps[nx][ny]:
                    visit[nx][ny] = visit[x][y] + 1
                    queue.append((nx, ny))
                else:
                    visit[nx][ny] = visit[x][y]
                    queue.appendleft((nx, ny))

n, m = map(int, input().split())
maps = [list(map(int, input().rstrip())) for _ in range(m)]
visit = [[-1] * n for _ in range(m)]
print(bfs())

File: 백준/Gold/1261. 알고스팟/알고스팟.java
import java.io.*;
import java.util.*;

class Main {
    static int n, m;
    static int [][] maps;
    static int [][] visit;
    static int [] dx = {0, 0, 1, -1};
    static int [] dy = {1, -1, 0, 0};

    public static class Point {
        int x, y;
        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static int BFS(Point start) {
        Deque<Point> queue = new LinkedList<Point>();
        queue.offer(start);
        visit[start.x][start.y] = 1;

        while (!queue.isEmpty()) {
            Point current = queue.poll();

            for (int i = 0; i < 4; i++) {
                int nx = current.x + dx[i];
                int ny = current.y + dy[i];

                if (0 <= nx && nx < n && 0 <= ny && ny < m && visit[nx][ny] == 0) {
                    if (maps[nx][ny] == 1) {
                        visit[nx][ny] = visit[current.x][current.y] + 1;
                        queue.offer(new Point(nx, ny));
                    }
                    else {
                        visit[nx][ny] = visit[current.x][current.y];
                        queue.offerFirst(new Point(nx, ny));
                    }
                }
            }
        }
        return visit[n - 1][m - 1] - 1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        String [] nums = br.readLine().split(" ");
        
        n = Integer.parseInt(nums[1]);
        m = Integer.parseInt(nums[0]);

        maps = new int [n][m];
        visit = new int [n][m];

        for (int i = 0; i < n; i++) {
            String [] row = br.readLine().split("");
            for (int j = 0; j < m; j++) {
                maps[i][j] = Integer.parseInt(row[j]);
            }
        }
        int ans = BFS(new Point(0, 0));
        System.out.println(ans);
    }
}